src/service/user-service.js

```diff
@@ -0,0 +1,161 @@
+const mongoSvc = require('./mongo-service');
+const User = require('../model/user-model');
+
+const QueryBuilder = function() {
+  const query = {}
+  return {
+    id: function(id) {
+      if (id) {
+        query['id'] = id
+      }
+      return this
+    },
+    name:function(name) {
+      if (name) {
+        query['name'] = name
+      }
+      return this
+    },
+    build: function() {
+      return query
+    }
+  }
+};
+
+const svc = {
+
+  /**
+   * retrieve users from db
+   * @param query {{ id: String, name: String }}
+   * @param option {{ page: number, limit: number }}
+   * @param logger { Pino }
+   * @return {{ docs: Array, total: number }}
+   */
+  getUser: async function(query, option, logger = log) {
+    const _log = logger.child({ module: 'user-service', method: 'getUser' });
+    _log.info('retrieve user, query: %j, option: %j', query, option);
+
+    const dbConnection = mongoSvc.getMongoConnection(_log);
+    if (dbConnection == null) {
+      _log.error('Cannot get mongodb connection.');
+      throw new errors.InternalServerError('Cannot get mongodb connection.');
+    }
+
+    try {
+      let q = QueryBuilder().id(query.id).name(query.name).build();
+      const userModel = User.getUserModel(dbConnection);
+      const { docs, total } = await userModel.paginate(q, option);
+      _log.info(`got total ${total} records.`);
+      _log.debug('docs: %j', docs);
+
+      return { docs, total };
+    } catch (err) {
+      _log.error('retrieve user error: %j', errors.fullStack(err));
+      throw new errors.InternalServerError(`Failed to retrieve user: ${err.message}`);
+    }
+  },
+
+  /**
+   * insert user doc to db
+   * @param body {{ id: String, name: String, dob: String, address: String, description: String }}
+   * @param logger { Pino }
+   * @return {{ docs: Array }}
+   */
+  createUser: async function(body, logger = log) {
+    const _log = logger.child({ module: 'user-service', method: 'createUser' });
+    _log.info('create user: %j', body);
+
+    const user = _.pick(body, ['id', 'name', 'dob', 'address', 'description']);
+    _.checkRequired(user, ['id', 'name']);
```

Comment:
is any reason to use lodash and hand written code for object validation instead of use mature library like joi/ajv/...?

Reply:
No special reason, haha.

I just thought this way is cleaner only for the field mandatory checking, no need a schema definition.
Of course, if we have the requirement of value validation, e.g. number checking, length checking, using joi library is a better choice.

Thanks for your suggestion. I will keep on enhancing my application.